libraries/AP_JSONParser/AP_JSONParser.cpp

```diff
@@ -0,0 +1,206 @@
+    memset(buf, '\0', st.st_size+1);  // keep valgrind happy
+    if (buf == nullptr) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "alloc failed %u", unsigned(st.st_size+1));
+        AP::FS().close(fd);
+        return false;
+    }
+
+    if (AP::FS().read(fd, buf, st.st_size) != st.st_size) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "failed to read json %s", filepath);
+        AP::FS().close(fd);
+        return false;
+    }
+    AP::FS().close(fd);
+
+    char *start = strchr(buf, '{');
+    if (!start) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "Invalid json %s", filepath);
+        return false;
+    }
+
+    /*
+      remove comments, as not allowed by the parser
+     */
+    for (char *p = strchr(buf,'#'); p; p=strchr(p+1, '#')) {
+        // clear to end of line
+        do {
+            *p++ = ' ';
+        } while (*p != '\n' && *p != '\r' && *p);
+    }
+
+    jsmn_init(&parser);
+
+    const int r = jsmn_parse(&parser,
+                             buf,
+                             st.st_size,
+                             t,
+                             t_count);
+    if (r < 0) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "Failed to parse %s", filepath);
+        return false;
+    }
+
+    num_used_tokens = uint16_t(r);
+
+    if (t[0].type != JSMN_OBJECT) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "malformed JSON %s", filepath);
+        return false;
+    }
+
+    token_offset = 0;
+
+    return true;
+}
+
+#endif
+
+bool AP_JSONParser::get_string_for_current_token(char *ret, uint8_t len) const
+{
+    const uint16_t json_text_len = t[token_offset].end - t[token_offset].start;
+    if (len <= json_text_len) {
+        return false;
+    }
+    strncpy(ret, &buf[t[token_offset].start], json_text_len);
+    ret[json_text_len] = 0;
+    return true;
+}
+
+bool AP_JSONParser::shift_token_string(char *ret, uint8_t len)
+{
+    if (token_offset >= num_used_tokens) {
+        return false;
+    }
+    if (t[token_offset].type != JSMN_STRING) {
+        return false;
+    }
+    if (!get_string_for_current_token(ret, len)) {
+        return false;
+    }
+    token_offset++;  // shift
+    return true;
+}
+
+bool AP_JSONParser::shift_token_float(float &ret)
+{
+    if (token_offset >= num_used_tokens) {
+        return false;
+    }
+    if (t[token_offset].type != JSMN_PRIMITIVE) {
+        return false;
+    }
+
+    char tmp_value[20];
+    if (!get_string_for_current_token(tmp_value, ARRAY_SIZE(tmp_value))) {
+        return false;
+    }
+
+    ret = atof(tmp_value);
+    token_offset++;
+    return true;
+}
+
+bool AP_JSONParser::shift_token_int32_t(int32_t &ret)
+{
+    if (token_offset >= num_used_tokens) {
+        return false;
+    }
+    if (t[token_offset].type != JSMN_PRIMITIVE) {
+        return false;
+    }
+
+    char tmp_value[20];
+    if (!get_string_for_current_token(tmp_value, ARRAY_SIZE(tmp_value))) {
+        return false;
+    }
+
+    ret = atoi(tmp_value);
```

Comment:
Do we ignore the `0` return failure here and then return `true`?

Reply:
We are?!

There are integer conversion functions that can return error here.  I don't think we use them ATM in AP and I'm not going to chase that :-)

Reply:
Thanks, just checking it was intentional. FYI, looks like `strtol` is already used in the codebase.